src/Routers/FilesRouter.js

```diff
@@ -87,41 +112,121 @@ export class FilesRouter {
       return;
     }
 
-    filesController
-      .createFile(config, filename, req.body, contentType)
-      .then(result => {
-        res.status(201);
-        res.set('Location', result.url);
-        res.json(result);
-      })
-      .catch(e => {
-        logger.error('Error creating a file: ', e);
-        next(
-          new Parse.Error(
-            Parse.Error.FILE_SAVE_ERROR,
-            `Could not store file: ${filename}.`
-          )
+    const base64 = req.body.toString('base64');
+    const file = new Parse.File(filename, { base64 }, contentType);
+    const { metadata = {}, tags = {} } = req.fileData || {};
+    file.setTags(tags);
+    file.setMetadata(metadata);
+    const fileSize = Buffer.byteLength(req.body);
+    const fileObject = { file, fileSize };
+    try {
+      // run beforeSaveFile trigger
+      const triggerResult = await triggers.maybeRunFileTrigger(
+        triggers.Types.beforeSaveFile,
+        fileObject,
+        config,
+        req.auth
+      )
+      let saveResult;
```

Comment:
I like all the PR but this logic we have here when the beforeSave trigger returns a file. What would be the usage case here? I am little bit afraid of its complexity and also to download the file to be changed. It can make the process to consume a lot of memory and possibly crash.

Reply:
The usage case that I initially wanted to use it for is as follows:

**Returning an already saved file:** This would allow me to create a hash with the file data that is trying to be saved and storing it in a "FileObject" collection in the database... When a new file is being saved I can compare the hash with other hashes in the collection and return an already saved file if I find a duplicate. (The project I'm working on uses a lot of thumbnails, many of which are the exact same across many users. This would allow me to not have to keep saving multiple copies of the same thumbnail files).

**Returning an unsaved file:** Since I am allowing a user to return a saved Parse.File, I also included the ability to return an unsaved Parse.File for completeness. This would allow a completely different file to be returned and saved in the event you dont want the originally uploaded file to be stored.

If you think that returning an unsaved Parse.File is not a good idea I could make it so that you can only return an already saved Parse.File? I figured giving the programmer the capability (maybe with a warning) would be ideal?

Reply:
I'm not sure. It's strange for me. Maybe a cloud code for this would be better? @dplewis any thoughts?

Reply:
Any thoughts on this?